packages/app-contracts/src/ABI.tsx

```diff
@@ -5,65 +5,177 @@
 import { I18nProps } from '@polkadot/ui-app/types';
 
 import React from 'react';
-import { InputFile } from '@polkadot/ui-app';
+import styled from 'styled-components';
+import { InputFile, Labelled, Messages } from '@polkadot/ui-app';
 import { ContractAbi } from '@polkadot/types';
 import { u8aToString } from '@polkadot/util';
 
 import translate from './translate';
 
 type Props = I18nProps & {
-  help: React.ReactNode,
+  contractAbi?: ContractAbi,
+  help?: React.ReactNode,
   isError?: boolean,
-  label: React.ReactNode,
-  onChange: (json: string | null, contractAbi: ContractAbi | null) => void
+  isDisabled?: boolean,
+  isRequired?: boolean,
+  label?: React.ReactNode,
+  onChange: (json: string | null, contractAbi: ContractAbi | null) => void,
+  onRemove?: () => void,
+  onRemoved?: () => void,
+  onSelect?: () => void
 };
 
 type State = {
-  abi?: Uint8Array | null,
+  contractAbi?: ContractAbi,
   isAbiValid: boolean,
-  name?: string,
-  placeholder?: React.ReactNode | null
+  isEmpty: boolean,
+  isError: boolean
 };
 
+const Normalize = styled.div`
+  min-height: 4rem;
+`;
+
 class ABI extends React.PureComponent<Props, State> {
   state: State = {
-    isAbiValid: true
+    isAbiValid: false,
+    isEmpty: true,
+    isError: false
   };
 
+  constructor (props: Props) {
+    super(props);
+    if (props.contractAbi) {
```

Comment:
(nit) extra blank line before

Suggestion -

```js
const isAbiValid = !!props.contractAbi;

      this.state = {
        contractAbi: props.contractAbi,
        isAbiValid,
        isEmpty: !isAbiValid,
        isError: isError: props.isError || (props.isRequired && !isAbiValid) || false
      };
```

- No extra ifs
- No ...this.state, we set all (so either outside the constructor, or in the constructor so all is together)